airflow/contrib/hooks/spark_submit_hook.py

```diff
@@ -262,16 +315,115 @@ def _process_log(self, itr):
+            if "driverState" in line:
+                self._driver_status = line.split(' : ')[1].replace(',', '').replace('\"', '').strip()
+
+            self.log.debug("spark driver status log: {}".format(line))
+
+    def _start_driver_state_tracking(self):
+        """
+        Polls the driver based on self._driver_id to get the state.
+        Finish successfully when the state is FINISHED.
+        Finish failed when the state is ERROR/UNKNOWN/KILLED/FAILED.
+
+        Possible states:
+            SUBMITTED: Submitted but not yet scheduled on a worker
+            RUNNING: Has been allocated to a worker to run
+            FINISHED: Previously ran and exited cleanly
+            RELAUNCHING: Exited non-zero or due to worker failure, but has not yet started running again
+            UNKNOWN: The state of the driver is temporarily not known due to master failure recovery
+            KILLED: A user manually killed this driver
+            FAILED: The driver exited non-zero and was not supervised
+            ERROR: Unable to run or restart due to an unrecoverable error (e.g. missing jar file)
+        """
+        # Keep polling as long as the driver is processing
+        while self._driver_status not in ["FINISHED", "UNKNOWN", "KILLED", "FAILED", "ERROR"]:
+
+            self.log.debug("polling state of spark driver with id {}".format(self._driver_id))
+
+            poll_drive_state_cmd = self._build_track_driver_state_command()
+            status_process = subprocess.Popen(poll_drive_state_cmd,
+                                              stdout=subprocess.PIPE,
+                                              stderr=subprocess.STDOUT,
+                                              bufsize=-1,
+                                              universal_newlines=True)
+
+            self._process_spark_status_log(iter(status_process.stdout.readline, ''))
+            returncode = self._sp.wait()
+
+            if returncode:
+                raise AirflowException(
+                    "Failed to poll for the driver state: returncode = {}".format(returncode)
+                    )
+
+            # Sleep for 1 second as we do not want to spam the cluster
+            time.sleep(1)
+
+    def _build_spark_driver_kill_command(self):
+        """
+        Construct the spark-submit command to kill a driver.
+
+        :return: full command to kill a driver
+        """
+
+        # If the spark_home is passed then build the spark-submit executable path using
+        # the spark_home; otherwise assume that spark-submit is present in the path to
+        # the executing user
+        if self._connection['spark_home']:
+            connection_cmd = [os.path.join(self._connection['spark_home'], 'bin', self._connection['spark_binary'])]
+        else:
+            connection_cmd = [self._connection['spark_binary']]
+
+        # The url ot the spark master
+        connection_cmd += ["--master", self._connection['master']]
```

Comment:
this if/else logic to build the beginning of the spark submit `connection_cmd` is duplicated in the code, I suggest we refactor it to have it in only place.